### PR TITLE
restore `extend` field in configuration 

### PR DESCRIPTION
**Pull request by OsmondJiang**

### CODE REVIEW

File: src/docfx/restore/RestoreMap.cs [hunk 1]
```diff
@@ -44,17 +49,23 @@ public bool TryGetUrlRestorePath(string remote, out string restorePath)
             return false;
         }
 
-        public string GetUrlRestorePath(string remote)
+        public string GetUrlRestorePath(string docsetPath, string path)
         {
-            Debug.Assert(!string.IsNullOrEmpty(remote));
+            Debug.Assert(!string.IsNullOrEmpty(path));
+
+            if (!HrefUtility.IsAbsoluteHref(path))
```
Comment (yufeih) [inline src/docfx/restore/RestoreMap.cs]: >IsAbsoluteHref [](start = 29, length = 14)

Can we have an method `HrefUtility.IsHttp`? `IsAbsolutePath` behaves differently between windows and linux. I was thinking about deprecating this method.

File: src/docfx/config/Config.cs [hunk 2]
```diff
@@ -193,68 +197,89 @@ private static void ValidateContributorConfig(ContributionConfig config, string
...
             }
 
             if (config == null)
                 config = new JObject();
-            if (!expand || !config.TryGetValue(ConfigConstants.Extend, out var objExtend))
+
+            if (!extend)
                 return config;
 
-            if (parents.Contains(configPath))
```
Comment (yufeih) [inline src/docfx/config/Config.cs]: what is the output of:

docfx.yml:
  extends: docfx.yml

Comment (OsmondJiang) [reply src/docfx/config/Config.cs]: just itself as we don't support deep level extends.

---
In reply to: [203657706](https://github.com/dotnet/docfx/pull/3101#discussion_r203657706) [](ancestors = 203657706)

Comment (OsmondJiang) [reply src/docfx/config/Config.cs]: i can add a test for that

---
In reply to: [203660745](https://github.com/dotnet/docfx/pull/3101#discussion_r203660745) [](ancestors = 203660745,203657706)

File: src/docfx/restore/RestoreMap.cs (hunk 1 above)
Comment (OsmondJiang) [reply src/docfx/restore/RestoreMap.cs]: add todo for that

---
In reply to: [203647447](https://github.com/dotnet/docfx/pull/3101#discussion_r203647447) [](ancestors = 203647447)

File: src/docfx/restore/Restore.cs [hunk 3]
```diff
@@ -13,13 +13,10 @@ namespace Microsoft.Docs.Build
 {
     internal static class Restore
     {
-        public static async Task Run(string docsetPath, CommandLineOptions options, Report report)
+        public static async Task Run(string docsetPath, CommandLineOptions options)
         {
             // Restore has to use Config directly, it cannot depend on Docset,
             // because Docset assumes the repo to physically exist on disk.
-            var config = Config.Load(docsetPath, options);
-            report.Configure(docsetPath, config);
```
[the remark this answers is not in the document]
Comment (OsmondJiang) [reply src/docfx/restore/Restore.cs]: I didn't see the reporter was used in restore, so i just remove it.

---
In reply to: [203658558](https://github.com/dotnet/docfx/pull/3101#discussion_r203658558) [](ancestors = 203658558)

File: src/docfx/config/Config.cs [hunk 4]
```diff
@@ -193,68 +197,89 @@ private static void ValidateContributorConfig(ContributionConfig config, string
...
+        private static JToken ExpandExtend(JToken extend)
+        {
+            if (extend == null)
+                return null;
+
+            if (extend is JValue extendValue)
+            {
+                extend = new JArray(extendValue);
+            }
+
```
Comment (superyyrrzz) [inline src/docfx/config/Config.cs]: can use `ExpandStringArray`

Comment (OsmondJiang) [reply src/docfx/config/Config.cs]: no error handling in that method but it still can be used. ;)

---
In reply to: [203667192](https://github.com/dotnet/docfx/pull/3101#discussion_r203667192) [](ancestors = 203667192)

**Pull request merged**